chat-plugins/poll.js

```diff
@@ -0,0 +1,158 @@
+/*
+* Poll chat plugin
+* This plugin allows roomauth (default: moderator and above) to run a poll in a room.
+* Every room can have one poll, and every user can vote. The results are displayed for
+* users that have voted, and are updated in real time. The poll can be closed with /endpoll.
+* By bumbadadabum with (a lot of) help from Zarel.
+*/
+
+var Poll = (function () {
+	function Poll(room, question, options) {
+		if (room.pollNumber) {
+			room.pollNumber++;
+		} else {
+			room.pollNumber = 1;
+		}
+		this.room = room;
+		this.question = question;
+		this.voters = new Set();
+		this.totalVotes = 0;
+
+		this.options = new Map();
+		for (var i = 0; i < options.length; i++) {
+			this.options.set(i + 1, {name: options[i], votes: 0});
+		}
+	}
+
+	Poll.prototype.vote = function (user, option) {
+		if (this.voters.has(user.latestIp)) {
+			return user.sendTo(this.room, "You have already voted for this poll.");
+		} else {
+			this.voters.add(user.latestIp);
+		}
+
+		this.options.get(option).votes++;
+		this.totalVotes++;
+
+		this.update();
+	};
+
+	Poll.prototype.generateVotes = function () {
+		var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+		this.options.forEach(function (option, number) {
+			output += '<button value="/votepoll ' + number + '" name="send">' + number + '. ' + Tools.escapeHTML(option.name) + '</button><br/>';
+		});
+		output += '</div>';
+
+		return output;
+	};
+
+	Poll.prototype.generateResults = function () {
+		var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+		var iter = this.options.entries();
+
+		var i = iter.next();
+		while (!i.done) {
+			output += '<span style="font-weight:bold;padding:2px">' + i.value[0] + '. ' + Tools.escapeHTML(i.value[1].name) + '</span><br/><span style="font-size:8pt;padding:0px">' + i.value[1].votes + ' votes (' + Math.round((i.value[1].votes * 100) / this.totalVotes) + '% of total).</span><br/>';
+			i = iter.next();
+		}
+		output += '</div>';
```

Comment:
Since there's already a client sided part for polls, don't generate HTML on the server side. Send the relevant data to the client and let it build the HTML to display